src/Compilers/CSharp/Portable/Compiler/Instrumentation.cs

```diff
@@ -0,0 +1,212 @@
+            newMethodBody = (BoundBlock)collector.Visit(methodBody);
+            dynamicAnalysisSpans = spansBuilder.ToImmutableAndFree();
+        }
+
+        private InstrumentationInjectionRewriter(MethodSymbol method, ArrayBuilder<SourceSpan> spansBuilder, FieldSymbol payload, TypeCompilationState compilationState, DiagnosticBag diagnostics, DebugDocumentProvider debugDocumentProvider)
+        {
+            _method = method;
+            _spansBuilder = spansBuilder;
+            _payload = payload;
+            _compilationState = compilationState;
+            _diagnostics = diagnostics;
+            _debugDocumentProvider = debugDocumentProvider;
+        }
+
+        public override BoundNode Visit(BoundNode node)
+        {
+            BoundStatement statement = node as BoundStatement;
+            if (statement != null)
+            {
+                switch (statement.Kind)
+                {
+                    case BoundKind.SwitchSection:
+                    case BoundKind.SwitchLabel:
+                    case BoundKind.PatternSwitchSection:
+                    case BoundKind.PatternSwitchLabel:
+                    case BoundKind.CatchBlock:
+                    // A labeled statement or a sequence point can be ignored with respect to instrumentation in favor of the underlying statement.
+                    case BoundKind.SequencePoint:
+                    case BoundKind.SequencePointWithSpan:
+                    case BoundKind.LabeledStatement:
+                        break;
+                    default:
+                        return CollectDynamicAnalysis(base.Visit(node));
+                }
+            }
+
+            return base.Visit(node);
+        }
+        
+        private BoundNode CollectDynamicAnalysis(BoundNode node)
+        {
+            BoundStatement statement = node as BoundStatement;
+            if (statement != null)
+            {
+                return CollectDynamicAnalysis(statement);
+            }
+
+            return node;
+        }
+
+        private BoundNode CollectDynamicAnalysis(BoundStatement statement)
+        {
+            if (statement.WasCompilerGenerated)
+            {
+                return statement;
+            }
+
+            Location statementLocation = statement.Syntax.GetLocation();
+            FileLinePositionSpan lineSpan = statementLocation.GetMappedLineSpan();
+            string path = lineSpan.Path;
+            if (path.Length == 0)
+            {
+                path = statement.Syntax.SyntaxTree.FilePath;
+            }
+
+            int spansIndex = _spansBuilder.Count;
+            _spansBuilder.Add(new SourceSpan(_debugDocumentProvider.Invoke(path, ""), lineSpan.StartLinePosition.Line, lineSpan.StartLinePosition.Character, lineSpan.EndLinePosition.Line, lineSpan.EndLinePosition.Character));
+
+            // Generate "_payload[pointIndex] = true".
```

Comment:
We should decide what behavior do we want under debugger. For example, when debugger stops at the statement in source, do we want this assignment happen, or do we want it to happen when user steps to execute the statement. I assume we would want the latter, because user might move instruction pointer to a different statement and resume execution. In this case we would do assignment for a statement that wasn't executed and likely wouldn't do it for the statement that would. So, it feels like we should do something special around sequence points. More reasons to move this code to LocalRewriter and coordinate with creation of sequence points.